pkg/scheduler/admission/podnodeconstraints/admission.go

```diff
@@ -0,0 +1,77 @@
+package podnodeconstraints
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"reflect"
+
+	configlatest "github.com/openshift/origin/pkg/cmd/server/api/latest"
+	"github.com/openshift/origin/pkg/scheduler/admission/podnodeconstraints/api"
+
+	admission "k8s.io/kubernetes/pkg/admission"
+	kapi "k8s.io/kubernetes/pkg/api"
+	kclient "k8s.io/kubernetes/pkg/client/unversioned"
+)
+
+func init() {
+	admission.RegisterPlugin("PodNodeConstraints", func(c kclient.Interface, config io.Reader) (admission.Interface, error) {
+		pluginConfig, err := readConfig(config)
+		if err != nil {
+			return nil, err
+		}
+		return NewPodNodeConstraints(pluginConfig), nil
+	})
+}
+
+func NewPodNodeConstraints(config *api.PodNodeConstraintsConfig) admission.Interface {
+	return &podNodeConstraints{
+		config:  config,
+		Handler: admission.NewHandler(admission.Create),
+	}
+}
+
+type podNodeConstraints struct {
+	*admission.Handler
+	config *api.PodNodeConstraintsConfig
+}
+
+func readConfig(reader io.Reader) (*api.PodNodeConstraintsConfig, error) {
+	if reader == nil || reflect.ValueOf(reader).IsNil() {
+		return &api.PodNodeConstraintsConfig{}, nil
+	}
+
+	configBytes, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return nil, err
+	}
+	config := &api.PodNodeConstraintsConfig{}
+	err = configlatest.ReadYAML(configBytes, config)
+	if err != nil {
+		return nil, err
+	}
+	// No validation needed since config is just 2 bools
+	return config, nil
+}
+
+func (oo *podNodeConstraints) Admit(aa admission.Attributes) error {
+	switch {
+	case oo.config == nil,
+		aa.GetResource() != kapi.Resource("pods"),
+		aa.GetKind().Kind != "Pod",
+		aa.GetSubresource() != "",
+		aa.GetOperation() != admission.Create:
+		return nil
+	}
+	pod, ok := aa.GetObject().(*kapi.Pod)
+	if !ok {
+		return admission.NewForbidden(aa, fmt.Errorf("unexpected object: %#v", aa.GetObject()))
+	}
+	if len(pod.Spec.NodeName) > 0 && oo.config.ProhibitNodeName {
```

Comment:
Whether it's SCC or not is irrelevant, I'm looking for a justification for
why we need this admission controller in light of the node labels.  I'd
like a concrete reason we *must have* the ability to limit node name,
couched in either cluster security or cluster stability terms, and then
address whether those issues should be solved elsewhere.

On Wed, Feb 17, 2016 at 2:05 PM, David Eads <notifications@github.com>
wrote:

> In pkg/scheduler/admission/podnodeconstraints/admission.go
> <https://github.com/openshift/origin/pull/6979#discussion_r53213589>:
>
> > +}
> > +
> > +func (oo *podNodeConstraints) Admit(aa admission.Attributes) error {
> > +	switch {
> > +	case oo.config == nil,
> > +		aa.GetResource() != kapi.Resource("pods"),
> > +		aa.GetKind().Kind != "Pod",
> > +		aa.GetSubresource() != "",
> > +		aa.GetOperation() != admission.Create:
> > +		return nil
> > +	}
> > +	pod, ok := aa.GetObject().(*kapi.Pod)
> > +	if !ok {
> > +		return admission.NewForbidden(aa, fmt.Errorf("unexpected object: %#v", aa.GetObject()))
> > +	}
> > +	if len(pod.Spec.NodeName) > 0 && oo.config.ProhibitNodeName {
>
> This is sounding more like pod security the more we talk about it....
>
> SCC is currently focused on what a particular pod can do, not on where a
> particular pod can live. I don't think it makes much sense to combine those
> concepts in this way. I could probably see: "a pod can do this on that
> node", but I really don't see SCC as something to control scheduling.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6979/files#r53213589>.
>